### ISSUE TITLE
Whois list microoptimalization suggestions :)

### ISSUE THREAD

**Issue by johnny-bit** [whois-server-list/whois-server-list#6]
Hi!

I'm currently writing a tool that consumes whois-server-list.xml for our for of php-whois. From my experience with this project and experience with _very enterprisey_ domain registration software, I'd like to make some suggestions :) In #1 you've said that you're open to changing schema and my suggestions would require that :wink:
- There are some whois servers that can ad are used for multiple TLDs, like `whois.donuts.co`. In case of updating availablePattern one would have to keep track of all 100+ occurrences of `whois.donuts.co`. Plus all those occurrences "puff up" xml file IMHO unnecessary. Would it be better to have separate list of whois servers and their patters/infos and in TLDs have links to them? Eg:

``` xml
<domainList>
    <domain name="esq">
        <source>IANA</source>
        <whoisServer>domain-registry-whois.l.google.com</whoisServer>
    </domain>
</domainList>
<whoisList>
    <whois host="domain-registry-whois.l.google.com">
        <availablePattern>\Qno match\E</availablePattern>
    </whois>
</whoisList>
```
- To build up upon 1st: some TLDs have multiple possible whois servers, all of which are able to give proper info about domain availability. Putting list of servers in TLD would help some implementations "cycle" requests to whois services. maybe something like this:

``` xml
<domainList>
    <domain name="sample">
        <source>IANA</source>
        <whoisServer>domain-registry-whois.l.google.com</whoisServer>
        <whoisServer>whois.donuts.co</whoisServer>
        <whoisServer>whois.sample.com</whoisServer>
        <whoisServer>whois.nic.sample</whoisServer>
    </domain>
</domainList>
```
- Many whois servers have requests limits or TOS requirements or straight up require pre-registrations to use them. It would be good to have that kind of info in whoisList :)
- Some registrars do not even have whois server responging on port 43. Instead relying on http-based service where one can do GET/POST request and receive appropriate info. I see that on list there are no http(s) based hosts. Would you consider including them or simply reject those as they are NOT by definition WHOIS servers?

Thanks for creating this list! It's way better to have centralised resource instead of doing "homebrew" for every project requiring whois info :)

**Comment by malkusch** [comment]
> In #1 you've said that you're open to changing schema

Yes, that was when the list was quite new. I'm still open to changes, but now that it is in use, I will consider breaking changes very carefully.

> In case of updating availablePattern one would have to keep track of all 100+ occurrences of whois.donuts.co. Plus all those occurrences "puff up" xml file IMHO unnecessary. Would it be better to have separate list of whois servers and their patters/infos and in TLDs have links to them?

Let's ignore puffing up the XML, as I consider this list as an intermediate format. Size should not matter much. But I can agree that updating whois server manually might become painfuly. The reason why it is like this, is flexibility: The script which generates the list, considers each top level domain as an isolated entity. I.e. it will try to figure out an availablePattern on each top level domain, no matter which availablePattern was found for a previous top level domain with the identical whois server. I do this on purpose, because I can't know for sure, that the pattern is identical. How can you be sure, that this case will never happen?

So I would trade here increased redundancy to gain flexibility. Besides that, I don't maintain the list manually, and the PRs don't suffer from the redundancy so far. The only benefit I can see here, would be less manual work, which didn't happen so far. I guess this is not worth breaking the format so hard.

> TLDs have multiple possible whois servers

That's already implemented in the [current schema](https://github.com/whois-server-list/whois-server-list/blob/2.2.4/whois-server-list.xsd#L32).

> Many whois servers have requests limits or TOS requirements or straight up require pre-registrations to use them. It would be good to have that kind of info in whoisList 

How would this information come into the list?

> Some registrars do not even have whois server responging on port 43. Instead relying on http-based service where one can do GET/POST request and receive appropriate info.

That's something what I'm definetly planing to integrate. It's worth discussing this isolated in #8.

**Comment by malkusch** [comment]
> I'm currently writing a tool that consumes whois-server-list.xml for our for of php-whois.

Oh, and let me use this opportunity for a commercial:You might be interested in using this existing [whois api](http://whois-api.domaininformation.de/) which also comes with a [PHP client](https://github.com/whois-server-list/whois-api-php).

**Comment by johnny-bit** [comment]
> Oh, and let me use this opportunity for a commercial:

Would be very interesting if my project wasn't limited now to single whois server to which I have 10k req/day :) 

> How would this information come into the list?

Volounteer info? :) Like for example situation with whois.nic.es, or difference registration makes in using whois.dns.pl

> I do this on purpose, because I can't know for sure, that the pattern is identical. How can you be sure, that this case will never happen?

Generic assumption for working with some registrars and clients for them - most work was defining whois servers and then attaching TLDs to them not other way around. But your way is good too - I do aproximation and use all availablePatterns for whois server found using regex grouping.
